File: src/api/mdns.ts
import { invoke } from '@tauri-apps/api/core';

/** Start the MDNS client */
export async function startMdnsClient(): Promise<void> {
  return invoke<void>('start_mdns_client');
}

/** Stop the MDNS client */
export async function stopMdnsClient(): Promise<void> {
  return invoke<void>('stop_mdns_client');
}

/** Start the MDNS server */
export async function startMdnsServer(): Promise<void> {
  return invoke<void>('start_mdns_server');
}

/** Stop the MDNS server */
export async function stopMdnsServer(): Promise<void> {
  return invoke<void>('stop_mdns_server');
}

export interface UpdateServerInfoOptions {
  host?: string;
  mdnsPort?: number;
  tcpPort?: number;
}

/** Update the MDNS server configuration */
export async function updateMdnsServerInfo(options: UpdateServerInfoOptions = {}): Promise<void> {
  return invoke<void>('update_mdns_server_info', {
    host: options.host,
    mdns_port: options.mdnsPort,
    tcp_port: options.tcpPort,
  });
}

File: src-tauri/src/api/mdns.rs
use spdlog::error;
use tauri::Runtime;

use crate::service::{client, server};

#[tauri::command]
pub async fn start_mdns_client() -> Result<(), String> {
    client::mdns::Mdns::instance().start().await.map_err(|e| {
        error!("Failed to start mdns client: {}", e);
        e.to_string()
    })
}

#[tauri::command]
pub async fn stop_mdns_client() -> Result<(), String> {
    client::mdns::Mdns::instance().stop().await.map_err(|e| {
        error!("Failed to stop mdns client: {}", e);
        e.to_string()
    })
}

#[tauri::command]
pub async fn start_mdns_server() -> Result<(), String> {
    server::mdns::Mdns::instance().start().await.map_err(|e| {
        error!("Failed to start mdns server: {}", e);
        e.to_string()
    })
}

#[tauri::command]
pub async fn stop_mdns_server() -> Result<(), String> {
    server::mdns::Mdns::instance().stop().await.map_err(|e| {
        error!("Failed to stop mdns server: {}", e);
        e.to_string()
    })
}

#[tauri::command]
pub async fn update_mdns_server_info<R: Runtime>(
    _app: tauri::AppHandle<R>,
    _window: tauri::Window<R>,
    host: Option<String>,
    mdns_port: Option<u16>,
    tcp_port: Option<u16>,
) -> Result<(), String> {
    server::mdns::Mdns::instance()
        .update_server_info(host, mdns_port, tcp_port)
        .await
        .map_err(|e| {
            error!("Failed to update mdns server info: {}", e);
            e.to_string()
        })
}
